### ISSUE TITLE
Precedence for ambiguous parsings?

### ISSUE THREAD

**Issue by beaugunderson** [Hardmath123/nearley#81]
Can't figure out how to change precedence for ambiguous parsings. My grammar is ambiguous because it includes emoji (a subset of unicode) and all unicode characters.

Because of this emoji can be parsed as a single emoji or as two constituent unicode characters--but I can't figure out how to prefer parsings where emoji are parsed as emoji (and not as their constituent characters).

**Comment by Hardmath123** [comment]
Ah, interesting question. Here's one way to do it:

```
text -> textchar:*
textchar -> ":-)" | [^:] | ":" [^-] | ":-" [^\)]
```

If your emojis are all exactly two unicode characters, then you have another option using the `reject` feature:

```
text -> digraph:+ .
digraph -> "=)" | "=(" | "=/" | . . {%
function(data, location, reject) {
  if (isAnEmoji(data.join("")) {
    return reject;
  } else {
    return data;
  }
}
%}
```

**Comment by beaugunderson** [comment]
My grammar looks like this:

```
@import "./emoji.pegjs" as Emoji

Unicode
 -> Symbol:*

Symbol
 -> Emoji
  | SurrogatePair
  | [\u0000-\uD799]

SurrogatePair
 -> [\uD800-\uDBFF][\uDC00-\uDFFF]

Emoji
 -> PhoneButton {% id %}
  | EmojiCharacter VariationSelector:? {%
      function (data, location, reject) {
        var e = data[0];
        var v = data[1];

        return (e.join ? e.join('') : e) + (v ? v.join ? v.join('') : v : '');
      }
    %}

PhoneButton
 -> [0-9#*] VariationSelector:? [\u20E3] {%
      function (data, location, reject) {
        return data.join('');
      }
    %}

EmojiCharacter
 -> [\uD83C] [\uDDE6-\uDDFF] [\uD83C] [\uDDE6-\uDDFF]
  | [\uD83C] [\uDC04-\uDFFF]
  | [\uD83D] [\uDC00-\uDE4F]
  | [\uD83D] [\uDE80-\uDEC5]
  | [\u203C-\u3299]
  | [\u24C2]
  | [\u2702-\u27B0]
  | [\u00A9\u00AE]

VariationSelector
 -> [\uFE0E\uFE0F]
```

The idea is that it can parse (or split) strings containing any unicode character, while being emoji-aware.

The flag emojis are actually four unicode characters long, the phone button emojis are three unicode characters long, some emojis are two unicode characters long (and so are ambiguous because alone they're each a valid unicode character).

So it seems like it is possible to add a regex for the 8 types of emoji constructions in `EmojiCharacter`, and use the `reject` but I'd like to avoid that if possible to keep things DRY. :)

Maybe just sorting the parsed outputs by least number of characters would suffice?

But I had hoped there was some other way (the bit on the website about returning all of the parsings in a deterministic order made me wonder what that order was and if I could set that order myself).

**Comment by Hardmath123** [comment]
Ah, I see.

Firstly, what's up with `@import "./emoji.pegjs" as Emoji`? Curious. :-)

I've been thinking, and I think I can hack in precedence if I think about it. The question is more about how the interface should be presented to the user. Where do you declare precedence? How? I hate Haskell's "fixity" operators. Should it be implicit in the order in which you declare the rules (which is, I think, how it works right now)? Or should you be able to assign ranks to production rules?

```
Math -> [1] Number "+" Number
      | [2] Number "*" Number
```

Lots of questions to answer. Let me know your opinion.

As it stands, though, this feature is probably a few days' work (and even more for writing rigorous test-cases!). Meanwhile, I would suggest going with your trick of picking the parsing with the fewest number of "characters".

---

(For posterity,) the cleaner `reject` way to do it is:

```
Unicode -> (Char Emoji):* Char
Emoji -> …
Char -> .:* {%
function(data, location, reject) {
  if (regex_test_for_emoji) {
    return reject;
  } else {
    return data;
  }
}
%}
```

**Comment by beaugunderson** [comment]
Oops, the `@import` line was a remnant from the PegJS version (which is achieved by using [pegjs-import](https://www.npmjs.com/package/pegjs-import) because PegJS does not natively support the `@import` statement). :)

I did try changing the order of the alternates but it didn't seem to affect the order of the found parsings (which made me think there was some other ordering happening).

Implicit declaration makes sense to me, but I also like how the explicit version looks so I don't have much of a preference. I'm not a parsing guru so I will defer to others here!

**Comment by tjvr** [comment]
@Hardmath123 Do you have any ideas for a syntax for this? :)

**Comment by Hardmath123** [comment]
I was thinking a strict arrow `-!>` that assigns the production rules separated by `|` precedence in the order they are specified.

```
char -!> ":-)" | .
```

(This example would be better served by a tokenizer, of course.)

**Comment by nanalan** [comment]
As an alternative to that arrow that doesn't require you to indent by another character 😉 
`=>`

**Comment by Hardmath123** [comment]
No. nearley already allows `=>` as a substitute for `->` (you can even do `a -===-----=> b` if you want). Backwards-compatibility matters.

**Comment by nanalan** [comment]
Whoops, didn't know that. Is that documented?

**Comment by Hardmath123** [comment]
No. </offtopic.>

**Comment by katywings** [comment]
Do I understand this thread correctly? The hypothetical `-!>` would fix this ambiguous grammar?

INPUT: "catsdasfdfasfeca
tsn23kj42342jk34nkcats"

```
MAIN -> CATSFTW
CATSFTW -> null
  | "cats" CATSFTW
  | [\s\S] CATSFTW
```

So the pipe `|` doesn't just take the first found correct pattern but *always* also goes through all other possibilities? I want to create a relatively small jsx, which only manipulates some special occurences in the input code. It works already pretty good, but my problem is that I have many [\s\S] cases... and therefore i get 30k different results even when I test a input file with 13 lines or so. This makes the parsing horrible slow xD.

Thanks a lot in advance!

**Comment by tjvr** [comment]
@katywings I don't think ordered choice aka. `-!>` is what you want there. You're probably better off restructuring your grammar not to be ambiguous! One way of doing this might be to use a tokeniser. (Also, I should note that in nearley, left recursion is slightly faster than right recursion.) If you need more help, open a fresh issue :-)

**Comment by Hardmath123** [comment]
(Also, I think `.` would match the same as `[\s\S]`.)

**Comment by katywings** [comment]
@tjvr Thanks for the tip with the left recursion, didnt know that. A tokenizer doesnt help because it would need the same complexity as my nearley grammar. Except from the `-!>` the only solution in nearley i can think of is to use *not*- Rules which would be kinda stupid:

```
CATIFY -> null
  | "cat" CATIFY
  | [\s\S] CATIFY {% (d, l, r) => {
if d startswith "cat" reject
}
%}
```

I don't think this is efficient because postprocessors get applied in the post process when I understand it correctly, therefore the not rule would need to go through all 30k different parser results.

So under the current circumstances I need to use an other parser library ;(.

@Hardmath123 Werent you the guy who somewhere in this thread used the word *offtopic*? :).  However: Thx I know... For prototyping I sometimes prefer the char class because the dot constantly gets ignored by my eyes ;)

**Comment by tjvr** [comment]
@katywings Can you show your grammar (on a new issue)? I don't understand what you're trying to do. Thanks.

**Comment by katywings** [comment]
@tjvr Thank you for your help. I can do that, probably at the weekend ;). 
Have a good night!


Am 18. April 2017 1:16:24 nachm. schrieb "Tim (blob8108)" 
<notifications@github.com>:
